### PR TITLE
Fix registry authorization errors

### LINKED ISSUE #26462: docker pull <unknown image> asks for auth on Windows
Running docker master on Windows

```
PS E:\> docker pull thisimagenamedoesnotexist
Using default tag: latest

Please login prior to pull:
Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.
Username: EOF
PS E:\>
```

I think (pretty sure) this is coming from the backend, so I'm not sure I can fix it in docker/docker itself. Doing the same against a Linux daemon results in `Error: image library/doesxnotxexist:latest not found`, which is more what I'd expect.

@dmcgowan @aaronlehmann Is this by design by any chance in the current hub backend?

### PR DESCRIPTION
**Pull request by dmcgowan**
Updates client to better handle errors when using OAuth tokens. Prevents unauthenticated errors getting sent back to the cli when permission is denied and erroneously prompting the user for login. This also trims error arrays down to a single element and logs the trimmed errors, this prevents the api from setting the status code to 401 when it sees "unauthorized" in one of the extra errors.

Needed for #28100
Closes #26462

### CODE REVIEW

**Comment by stevvooe** [comment]
LGTM

**Comment by thaJeztah** [comment]
@dmcgowan tests are failing

File: distribution/errors.go [hunk 1]
```diff
@@ -56,6 +59,37 @@ func shouldV2Fallback(err errcode.Error) bool {
...
+			}
+			return translatePullError(v[0], ref)
+		}
+	case errcode.Error:
+		var newErr error
+		switch v.Code {
+		case errcode.ErrorCodeDenied:
+			// ErrorCodeDenied is used when access to the repository was denied
+			newErr = errors.Errorf("repository %s not found: does not exist or no read access", ref.Name())
+		case v2.ErrorCodeManifestUnknown:
```
Comment (dmcgowan) [inline distribution/errors.go]: This is getting called when the error is terminal and will be returned back to the api handler. Right now the message getting returned by unauthorized is "correctly" getting picked up and prompting login. I am not too concerned with handling that case but after this PR I want to enumerate what all the cases are and the messaging getting sent to users.

**Comment by dmcgowan** [comment]
Looks like trust failures are related to https://github.com/docker/distribution/pull/2044 in distribution. Going to update the trust build code to ensure the output is not changing due to this change.

**Comment by dmcgowan** [comment]
Added a commit which resolves the issues with the distribution vendor update. The change to `WithDigest` required an additional method to remove tag before calling to preserve existing behavior. We have already discussed this on the distribution issue as the correct way to use `WithDigest` rather than assuming it will remove the tag.

**Comment by aaronlehmann** [comment]
Tested `docker pull` and `docker service create` with `name:tag@digest` syntax and didn't find any problems.

LGTM

**Comment by aaronlehmann** [comment]
However it's very possible that the only reason why these things work correctly is that `github.com/docker/docker/reference` drops the tag in `ParseNamed`:

```
        if canonical, isCanonical := named.(distreference.Canonical); isCanonical {
                return WithDigest(r, canonical.Digest())
        }
        if tagged, isTagged := named.(distreference.NamedTagged); isTagged {
                return WithTag(r, tagged.Tag())
        }
```

We should probably fix `ParseNamed` to be consistent with the upstream changes. But that may cause more issues.

**Comment by dmcgowan** [comment]
@aaronlehmann I saw that in `ParseNamed` figuring out where to update. I recall that was done intentionally to prevent the creation of a tagged and digested reference, but there is no comment explaining it and I can't think of an error it would cause. Seems to me like more of a problem to have it strip data such that `ParsedNamed(s).String() != s`. We can address this upon deprecation of the `docker/docker/reference` fork.

**Comment by aaronlehmann** [comment]
@dmcgowan: It turns out this is an issue for `service create`. Creating a service with the image reference `busybox:latest@sha256:29f5d56d12684887bdfa50dcd29fc31eea4aaf4ad3bec43daf19026a7ce69912` strips out the tag in `service ls`, whereas specifying `busybox:latest` keeps the tag and adds a digest as well. I suggested using `ParseNamed from `github.com/docker/distribution/reference` to work around this.

cc @nishanttotla

**Comment by tonistiigi** [comment]
LGTM

**Comment by vieux** [comment]
let's wait for https://github.com/docker/docker/pull/27917 before merging

**Comment by nishanttotla** [comment]
Adding some utility functions so that we can move to using `ParseNamed` from `docker/distribution/reference` here: https://github.com/docker/distribution/pull/2062

**Pull request merged**